src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +65,740 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+        private static List<Project> GetOrderedProjectsToExamine(
+            Solution solution,
+            IImmutableSet<Project> projects,
+            ProjectDependencyGraph dependencyGraph,
+            IEnumerable<ProjectId> projectsThatCouldReferenceType)
         {
-            dependentTypeIds = null;
+            var projectsToExamine = GetProjectsToExamineWorker(
+                solution, projects, dependencyGraph, projectsThatCouldReferenceType);
 
-            ConcurrentDictionary<SymbolKey, List<SymbolKey>> dictionary;
-
-            return cache.TryGetValue(compilation, out dictionary) &&
-                   dictionary.TryGetValue(typeId, out dependentTypeIds);
+            // Ensure the projects we're going to examine are ordered topologically.
+            // That way we can just sweep over them from left to right as no project
+            // could affect a previous project in the sweep.
+            return OrderTopologically(dependencyGraph, projectsToExamine);
         }
 
-        private static List<SymbolKey> GetOrAddDependentTypes(ConditionalWeakTable<Compilation, ConcurrentDictionary<SymbolKey, List<SymbolKey>>> cache, Compilation compilation, SymbolKey typeId, List<SymbolKey> dependentTypeIds)
+        private static List<Project> OrderTopologically(
+            ProjectDependencyGraph dependencyGraph, IEnumerable<Project> projectsToExamine)
         {
-            List<SymbolKey> result;
-            if (TryGetDependentTypes(cache, compilation, typeId, out result))
-            {
-                return result;
-            }
-            else
+            var order = new Dictionary<ProjectId, int>();
+
+            int index = 0;
+            foreach (var projectId in dependencyGraph.GetTopologicallySortedProjects())
             {
-                return cache.GetValue(compilation, s_createSymbolDictionary)
-                            .GetOrAdd(typeId, dependentTypeIds);
+                order.Add(projectId, index);
+                index++;
             }
+
+            return projectsToExamine.OrderBy((p1, p2) => order[p1.Id] - order[p2.Id]).ToList();
         }
 
-        internal static bool OriginalSymbolsMatch(
-            ISymbol searchSymbol,
-            ISymbol symbolToMatch,
+        private static IEnumerable<Project> GetProjectsToExamineWorker(
             Solution solution,
-            CancellationToken cancellationToken)
+            IImmutableSet<Project> projects,
+            ProjectDependencyGraph dependencyGraph,
+            IEnumerable<ProjectId> projectsThatCouldReferenceType)
         {
-            if (ReferenceEquals(searchSymbol, symbolToMatch))
+            // We need to search all projects that could reference the type *and* which are 
+            // referenced by some project in the project set.
+            foreach (var projectThatCouldReferenceType in projectsThatCouldReferenceType)
```

Comment:
I'll see if i can figure out how.  Note: this did not seem to be a problem in practice, even in a huge solution like roslyn.